improver/regrid/grid.py

```diff
@@ -0,0 +1,401 @@
+    return out_classified
+
+
+def classify_input_surface_type(cube_in_mask, classify_latlons):
+    """
+    Classify surface types of source grid points based on a binary True/False land mask
+    cube_in_mask's grid could be different from input source grid of NWP results
+
+    Args:
+        cube_in_mask (iris.cube.Cube):
+            land_sea mask information cube for input source grid(land=>1)
+            should in GeogCS's lats/lons coordinate system
+        classify_latlons(numpy.ndarray):
+            latitude and longitude source grid points to classify (N x 2)
+
+    Returns:
+        numpy.ndarray: classifications (N) for 1D-ordered source grid points
+    """
+    in_land_mask = cube_in_mask.data
+    lats_name, lons_name = latlon_names(cube_in_mask)
+    in_land_mask_lats = cube_in_mask.coord(lats_name).points
+    in_land_mask_lons = cube_in_mask.coord(lons_name).points
+
+    mask_rg_interp = RegularGridInterpolator(
+        (in_land_mask_lats, in_land_mask_lons),
+        in_land_mask,
+        method="nearest",
+        bounds_error=False,
+        fill_value=0.0,
+    )
+    is_land = np.bool_(mask_rg_interp(classify_latlons))
+    return is_land
+
+
+def slice_cube_by_domain(cube_in, output_domain):
+    """
+    extract cube domain to be consistent as cube_reference's domain
+
+    Args:
+        cube_in (iris.cube.Cube):
+            input data cube to be sliced
+        output_domain(tuple):
+            lat_max,lon_max,lat_min,lon_min
+
+    Returns:
+        cube_in (iris.cube.Cube):
+            data cube after slicing
+    """
+    lat_max, lon_max, lat_min, lon_min = output_domain
+    lat_d, lon_d = get_grid_size(cube_in)
+
+    domain = iris.Constraint(
+        latitude=lambda val: lat_min - 2.0 * lat_d < val < lat_max + 2.0 * lat_d
+    ) & iris.Constraint(
+        longitude=lambda val: lon_min - 2.0 * lon_d < val < lon_max + 2.0 * lon_d
+    )
+
+    cube_in = cube_in.extract(domain)
+
+    return cube_in
+
+
+def slice_mask_cube_by_domain(cube_in, cube_in_mask, output_domain):
+    """
+    extract cube domain to be consistent as cube_reference's domain
+
+    Args:
+        cube_in (iris.cube.Cube):
+            input data cube to be sliced
+        cube_in_mask (iris.cube.Cube):
+            input maskcube to be sliced
+        output_domain (Tuple[float, float, float, float]):
+            lat_max, lon_max, lat_min, lon_min
+    Returns:
+        Tuple[iris.cube.cube, iris.cube.Cube]:
+            data cube after slicing, mask cube after slicing
+    """
+    lat_max, lon_max, lat_min, lon_min = output_domain
+    lat_d_1, lon_d_1 = get_grid_size(cube_in)
+    lat_d_2, lon_d_2 = get_grid_size(cube_in_mask)
+    lat_d = lat_d_1 if lat_d_1 > lat_d_2 else lat_d_2
+    lon_d = lon_d_1 if lon_d_1 > lon_d_2 else lon_d_2
+
+    domain = iris.Constraint(
+        latitude=lambda val: lat_min - 2.0 * lat_d < val < lat_max + 2.0 * lat_d
```

Comment:
What the reason for choosing a domain this big to cut out? I agree you need a bit of a border on it, but twice the size of the domain seems quite large.

 Also shouldn't cube_in and cube_in_mask always be the same shape and size? We don't currently check this but from I think this should always be true, can you think of a reason why it wouldn't be true. I wonder if we did check that they are on the same domain much higher up in the code we could simplify some of your code. 

We also have two different utilities that can both be used to extract subsets of cubes. I'm not sure if either could replace this code, I'll have a think. (The two utilities are: https://github.com/metoppv/improver/blob/3cfbf3323d3a693a9c61fec13350295b85d03676/improver/utilities/cube_extraction.py#L228 and https://github.com/metoppv/improver/blob/3cfbf3323d3a693a9c61fec13350295b85d03676/improver/utilities/cube_extraction.py#L261

Reply:
The first code is to add just one grid size, but Iris slice failed one test case (not worked as I expected). I  changed it to twice.
 
cube_in and  cube_in_mask could be different in old BOM's code, and therefore we keep this flexible way.